extensions/amp-live-list/0.1/live-list-manager.js

```diff
@@ -0,0 +1,171 @@
+/**
+ * Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {Poller} from './poller';
+import {getService} from '../../../src/service';
+import {user} from '../../../src/log';
+import {viewerFor} from '../../../src/viewer';
+import {whenDocumentReady} from '../../../src/document-ready';
+import {xhrFor} from '../../../src/xhr';
+
+
+/**
+ * Manages registered AmpLiveList components.
+ * Primarily handles network requests and updates the components
+ * if necessary.
+ *
+ * @visibleForTesting
+ */
+export class LiveListManager {
+
+  constructor(win) {
+    this.win = win;
+
+    /** @private @const {!Object<string, !AmpLiveList>} */
+    this.liveLists_ = Object.create(null);
+
+    /** @private @const {!Viewer} */
+    this.viewer_ = viewerFor(this.win);
+
+    /** @private {number} */
+    this.interval_ = 15000;
+
+    /** @private @const {!Array<number>} */
+    this.intervals_ = [this.interval_];
+
+    /** @private {?Poller} */
+    this.poller_ = null;
+
+    /** @private @const {string} */
+    this.url_ = this.win.location.href;
+
+    /** @private @const {function(): Promise} */
+    this.work_ = this.fetchDocument_.bind(this, this.url_);
+
+    // Only start polling when doc is ready and when the viewer is visible.
+    this.whenDocReady_().then(() => {
+      // Switch out the poller interval if we can find a lower one and
+      // then make sure to stop polling if viewer is not visible.
+      this.interval_ = Math.min.apply(Math, this.intervals_);
+      this.poller_ = new Poller(this.win, this.interval_, this.work_);
+
+      if (this.viewer_.isVisible()) {
+        this.poller_.start();
+      }
+      this.setupVisibilityHandler_();
+    });
+  }
+
+  /**
+   * Makes a request to the given url for the latest document.
+   *
+   * @param {string} url
+   * @private
+   */
+  fetchDocument_(url) {
+    return xhrFor(this.win)
+        // TODO(erwinm): add update time here when possible.
+        .fetchDocument(url)
+        .then(this.getLiveLists_.bind(this));
+  }
+
+  /**
+   * Extract the `Document` instance parsed through responseType=document
+   * from the default xhr parser. This is important because we dont want
+   * to do any parsing in the UI thread.
+   *
+   * @param {!FetchResponse} res
+   * @return {!HTMLDocument}
+   */
+  getDoc(res) {
+    const doc = res.doc();
+    user.assert(doc instanceof Document,
```

Comment:
`instanceof` is dangerous, because it must come from the correct window. This might be always correct if we get the XHR object from the main window as well.